File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:


int trap(vector<int>& heights) {
    int len = heights.size();
    if (len == 0) return 0; // Edge case: empty input

    // Preallocate left max vector
    vector<int> mx_vec_left(len);
    int mx = heights[0];
    for (int i = 0; i < len; i++) {
        mx = max(mx, heights[i]);
        mx_vec_left[i] = mx;
    }

    // Preallocate right max vector
    vector<int> mx_vec_right(len);
    mx = heights[len - 1];

    // Fill right max from back to front
    for (int i = len - 1; i >= 0; i--) {
        mx = max(mx, heights[i]);
        mx_vec_right[i] = mx;
    }

    int ans = 0;

    // Compute trapped water
    for (int i = 0; i < len; i++) {
        int temp = min(mx_vec_left[i], mx_vec_right[i]) - heights[i];
        if (temp > 0) {
            ans += temp;
        }
    }

    return ans;
}

};

File: Data Structures & Algorithms/max-water-container/submission-1.py


class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l, r, heights_len = 0, 1, len(heights)
        ans = 0
        while r < heights_len:
            min_height = min(heights[l], heights[r])
            area = min_height * (r - l)
            ans = max(ans, area)
            if heights[r] > heights[l]:
                l = r
            r += 1

        return ans
